examples/06-fetch-map/legend.ts

```diff
@@ -0,0 +1,107 @@
+    layerDiv.appendChild(nameDiv);
+
+    const columnDiv = document.createElement('div');
+    columnDiv.className = 'legend-header';
+    columnDiv.textContent = `COLOR BASED ON`;
+    layerDiv.appendChild(columnDiv);
+
+    const dataColumnDiv = document.createElement('div');
+    dataColumnDiv.className = 'legend-column';
+    dataColumnDiv.textContent = dataColumn;
+    layerDiv.appendChild(dataColumnDiv);
+
+    const tilestats = layer.props.data?.tilestats?.layers[0]?.attributes?.find(
+      (a: any) => a.attribute === dataColumn
+    );
+
+    const isConstantColor = typeof layer.props.getFillColor !== 'function';
+
+    if (isConstantColor) {
+      const rangeDiv = document.createElement('div');
+      rangeDiv.className = 'legend-range';
+
+      const colorSwatch = document.createElement('div');
+      colorSwatch.className = 'legend-color-swatch';
+      const color = layer.props.getFillColor;
+      colorSwatch.style.backgroundColor = `rgb(${color[0]}, ${color[1]}, ${color[2]})`;
+
+      const rangeLabel = document.createElement('div');
+      rangeLabel.className = 'legend-range-label';
+      rangeLabel.textContent = 'All values';
+
+      rangeDiv.appendChild(colorSwatch);
+      rangeDiv.appendChild(rangeLabel);
+      layerDiv.appendChild(rangeDiv);
+    } else if (
+      tilestats &&
+      tilestats.type === 'Number' &&
+      tilestats.min !== undefined
+    ) {
+      const numRanges = 6;
+      const min = tilestats.min;
+      const max = tilestats.max;
+      const step = (max - min) / numRanges;
+
+      for (let i = 0; i < numRanges; i++) {
+        const rangeStart = min + step * i;
+        const rangeEnd = i === numRanges - 1 ? max : min + step * (i + 1);
+
+        const rangeDiv = document.createElement('div');
+        rangeDiv.className = 'legend-range';
+
+        const colorSwatch = document.createElement('div');
+        colorSwatch.className = 'legend-color-swatch';
+
+        const value = (rangeStart + rangeEnd) / 2;
+        const color = layer.props.getFillColor({
+          properties: {[dataColumn]: value},
+        });
+        colorSwatch.style.backgroundColor = `rgb(${color[0]}, ${color[1]}, ${color[2]})`;
+
+        const rangeLabel = document.createElement('div');
+        rangeLabel.className = 'legend-range-label';
+        rangeLabel.textContent = `${rangeStart.toFixed(2)} – ${rangeEnd.toFixed(2)}`;
+
+        rangeDiv.appendChild(colorSwatch);
+        rangeDiv.appendChild(rangeLabel);
+        layerDiv.appendChild(rangeDiv);
+      }
+    } else if (tilestats?.type === 'String' && tilestats.categories) {
+      const categoriesDiv = document.createElement('div');
+      categoriesDiv.className = 'legend-categories';
+      categoriesDiv.textContent = `Categories: ${tilestats.categories.length}`;
+      layerDiv.appendChild(categoriesDiv);
+    }
+
+    container.appendChild(layerDiv);
+  });
+
+  wrapper.appendChild(container);
+  return wrapper;
+}
```

Comment:
Optional – I think constructing the HTML programmatically is completely OK since we're trying to avoid  dependencies on frameworks here. If you did prefer something more ergonomic for this or future examples though, adding a lightweight HTML templater like [lit-html](https://www.npmjs.com/package/lit-html) would also be fine with me. (or others, mainly suggesting that because I've used it before and it's nearly dependency-free).

Reply:
I like your philosophy of not tying api-client to any frontend framework, so I went in the spirit of that and avoiding adding a dependency. For such a small example I think it is OK to just use vanilla JS